pgx/src/datum/from.rs

```diff
@@ -50,29 +70,48 @@ pub trait FromDatum {
     ///
     /// ## Safety
     ///
-    /// Same caveats as `From::from_datum(...)`
+    /// Same caveats as `FromDatum::from_datum(...)`
     unsafe fn from_datum_in_memory_context(
         mut memory_context: PgMemoryContexts,
         datum: pg_sys::Datum,
         is_null: bool,
-        typoid: pg_sys::Oid,
     ) -> Option<Self>
     where
         Self: Sized,
     {
-        memory_context.switch_to(|_| FromDatum::from_datum(datum, is_null, typoid))
+        memory_context.switch_to(|_| FromDatum::from_datum(datum, is_null))
+    }
+
+    /// `try_from_datum` is a convenience wrapper around `FromDatum::from_datum` that returns a
```

Comment:
Is there any reason to use `from_datum` now, given that `try_from_datum` exists? Perhaps the documentation should suggest `try_from_datum` as the primary API, not `from_datum`.

Reply:
In most cases, `from_datum` is preferred because we know the types are compatible -- most especially for the args to `#[pg_extern]` functions.  Additionally, there's less overhead in `from_datum` as we don't have to go digging for the Datum's type from $somewhere.

This was actually a thing @JLockerman once complained about and is how we ended up with that, now removed, `const NEEDS_TYPID: bool = false` thing in `FromDatum`.

Reply:
Hmm ok cool!

Reply:
Places where we don't know the types are when dealing with composite/row types, random HeapTuples, and technically with Spi.

So once this is merged, I'll plumb the usage of `try_from_datum` into Spi and also wrap up the composite type support.